source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,161 @@ namespace Slang
+            // a parent and a child to both have `__init()` constructors.
+            // As a result, this can cause a compile error when relying on the default
+            // constructor of an object because a child can use a parent object's
+            // constructor.
+            // To solve this issue and produce C++ like behavior we need to add overload
+            // ranks if not present already to prioritize the default constructor of a child.
+            // If an overload rank is manually set, we cannot assume any preference of
+            // constructor by the user and should ignore any special logic.
+            if (ctorDefault->findModifier<OverloadRankAttribute>())
+                return;
+            auto overloadRank = m_astBuilder->create<OverloadRankAttribute>();
+            overloadRank->rank = -levelsOfInheritance;
+            addModifier(ctorDefault, overloadRank);
+        };
+
+        // To follow C++ constructor behavior when adding a init expression to a member
+        // variable we must insert init expressions to the start of all constructors.
+        // Furthermore, to follow C++ behavior we must also ensure we use init expressions
+        // of all accessable parent member variables.
+        Dictionary<Decl*, Expr*> cachedDeclToCheckedVar;
+        for (auto aggTypeDecl : linearInheritanceHierarchy)
+        for (auto m : aggTypeDecl->members)
+        {
+            auto varDeclBase = as<VarDeclBase>(m);
+            if (varDeclBase
+                && varDeclBase->initExpr)
+            {
+                if (!fillCtorList)
+                    findAllCtor();
+
+                for (auto ctor : ctorList)
+                {
+                    auto body = as<BlockStmt>(ctor->body);
+                    auto seqStmt = as<SeqStmt>(body->body);
+
+                    VarExpr* memberVarExpr = m_astBuilder->create<VarExpr>();
+                    memberVarExpr->scope = ctor->ownedScope;
+                    memberVarExpr->name = m->getName();
+                    memberVarExpr->type = nullptr;
+                    auto assign = m_astBuilder->create<AssignExpr>();
+                    assign->left = memberVarExpr;
+                    assign->right = varDeclBase->initExpr;
+                    auto stmt = m_astBuilder->create<ExpressionStmt>();
+                    stmt->expression = assign;
+
+                    Expr* checkedMemberVarExpr;
+                    if (cachedDeclToCheckedVar.containsKey(m))
+                        checkedMemberVarExpr = cachedDeclToCheckedVar[m];
+                    else
+                    {
+                        checkedMemberVarExpr = CheckTerm(memberVarExpr);
+                        cachedDeclToCheckedVar.add({ m, checkedMemberVarExpr });
+                    }
+                    if (!checkedMemberVarExpr->type.isLeftValue)
+                        continue;
+
+                    if (seqStmt->stmts.getCount() == 0)
```

Comment:
Do you have to check the getCount()?
Can't we just call `seqStmt->stmts.insert(0, stmt);`?